### DISCUSSION TITLE
Besoin d'aide

### DISCUSSION THREAD

**Discussion by Fa9012** [iptv-org/community#552] [Q&A]
Bonjour, je vais m'exprimer en français puisque je en Afrique francophone. 
J'apprécie ce que vous faites et je vous remercie mais voilà j'utilise un décodeur qui ne prend que xtream ( url, username et mot de passe) alors j'aimerais vous suggérer de tenir compte pour nous permettre d'en bénéficier. Il est impossible d'insérer l'URL dans mon décodeur. Merci

**Comment by BellezaEmporium** [comment]
Nous ne proposons pas notre contenu sur Xtream-Codes. Nous n'en acceptons d'ailleurs pas sur nos listes, étant jugées trop peu stables (et limitées en nombre de personnes simultanées).

**Comment by rickeymandraque** [comment]
Bonjour ( et pavé César ! )
# TL;DR :
Si c'est réellement un décodeur avec une parabole/antenne au bout, il n'y a pas de solution et il n'y en aura jamais même si ce décodeur est relié au net (type dreambox).
Si c'est un démultiplexeur/boitier audiovisuel type Android TV ( un ordinateur en gros ! ) veuillez lire la suite, il existe peut-être des solutions tordues.

# Explication :
la seule solution que je vois (autre que de développer une application serveur spécifique imitant Xtream bien sûr) serait de créer vous-même un faux serveur Xtream avec des redirections vers les bon fichiers json, M3U ; m3u8 et XML (epg) avec un login factice sur un serveur gratuit type .free.fr ou tout autre service (ne le faites, c'est juste un exemple ! ) qui permet d'afficher des pages web, avec redirections, gratuitement ou peu cher.

Je travaille dessus justement et la tâche n'est pas aisée.

Il faudrait rediriger chaque requête `player_api.php`/`panel_api.php` vers le bon fichier json, les bonnes requêtes `get.php`/`xmltv.php`vers les bons fichiers XML ; M3U et m3u8...

(Ce qui induit : chaque catégories, chaques chaîne, bref , chaque cas de figure.)


Et pour ce qui concerne les flux en eux-mêmes, il faut créer un fichier json avec les flux  de iptv-org dans la valeur de la clef `"direct_source":` pour leurrer le lecteur et qu'il ne cherche pas à obtenir le flux du serveur en lui-même car il ne sera pas en mesure de transcoder le flux à la volée.
Ou bien de créer un fichier .m3u8 contenant le lien du flux avec le bon identifiant, ce qui revient au même que direct_source et ce qui veut dire 1fichier= 1 chaîne ! ( Donc une redirection supplémentaire).
Si vous partez 9 langues et que vous avez 900 chaînes... Bonjour l'angoisse !
Et encore, ce n'est pas dit que cela fonctionne correctement. ( C'est purement hypothétique)


```json
"11": { 
            "num": 11, 
            "name": "TF1",
            "stream_type": "live",
            "type_name": "Live Streams", 
            "stream_id": "11.m3u8",
            "stream_icon": "http://exemple.com/tf1.png",
            "epg_channel_id": "TF1.fr",
            "added": "1684788441", 
            "category_name": "FRANCE TNT", 
            "category_id": "1", 
            "series_no": null,
            "live": "1", 
            "container_extension": null,
            "custom_sid": null,
            "tv_archive": 0, 
            "direct_source": "https://exemple.com/playlist.m3u8",
            "tv_archive_duration": 0
        },
```

Toutes ces pages seraient statiques mais étant donné que le lecteur se sert de variables et de constantes pour afficher les chaînes en json/M3U sans jamais utiliser une vraie page PHP, ça pourrait faire le travail.
La mise à jour des chaînes et des fichiers pourrait se faire par script à grand renfort de curl, jq, grep, awk et autres joyeusetés du genre avec les différents outils de iptv-org (database, api, ... )

Là où cela se complique c'est le nombre de fichiers à gérer, à générer et à rediriger pour chaque requêtes du lecteur.
Mais c'est faisable avec un login connus à l'avance (exemple : fakeuser et Pa$$w0rd).

Voilà ce que pourrait donner l'arborescence d'un tel projet :
(X étant le nom d'utilisateur, on génère tout ça depuis Github et il resterait à envoyer tout ceci vers un serveur supportant les redirections)
```lua

/ (racine)
|-- .github/ (dossier pour les fichiers GitHub Actions)
|   |-- workflows/
|       |-- update_stats.yml (workflow pour mettre à jour les stats)
|-- scripts/ (dossier pour les scripts)
|   |-- generate_json.sh (script pour générer les fichiers JSON)
|   |-- generate_epg_xml.php (script pour générer l'EPG en XML)
|-- public_info/ (dossier pour les fichiers JSON d'information utilisateur)
|   |-- public.json
|-- server_info/ (dossier pour les fichiers JSON d'information serveur)
|   |-- server.json
|-- categories/ (dossier pour les catégories)
|   |-- live_categories.json
|   |-- vod_categories.json
|   |-- series_categories.json
|-- live_streams/ (dossier pour les flux en direct)
|   |-- all_live_streams.json
|   |-- category_X_live_streams.json
|-- vod_streams/ (dossier pour les flux VOD)
|   |-- all_vod_streams.json
|   |-- category_X_vod_streams.json
|-- series/ (dossier pour les séries)
|   |-- all_series.json
|   |-- category_X_series.json
|-- series_info/ (dossier pour les informations sur les séries)
|   |-- series_X_info.json
|-- vod_info/ (dossier pour les informations sur les VOD)
|   |-- vod_X_info.json
|-- epg/ (dossier pour les EPG)
|   |-- short_epg/ (dossier pour les courts EPG)
|       |-- stream_X_short_epg.json
|   |-- full_epg.xml
|-- .htaccess
|-- redirect.php
```
Et là, il n'y a pas les flux !

Avec cette solution, on dit adieu à tout ce qui est dynamique (nombre de connexion actuelle par exemple) et à la sécurité.

Surtout si un tel "serveur" est rendu publique et qu'il n'est pas capable de gérer les utilisateurs, la sécurité, etc, il va être pris d'assaut et le prestataire ne va pas être très content même si ce n'est que des petits fichiers de quelques Ko à chaque requête.

L'autre solution serait de créer soi-même un vrai serveur Xtream personnel sur un serveur dédié distant ou en locale (réel, virtuel ou en conteneur), mais là ce n'est pas le même prix, pas les mêmes compétences et pas les mêmes risques que se soit niveau sécurité ou légal.
Après si vous êtes prêt à dépenser 500$/mois pour ça ou utiliser une machine dédiée avec une carte graphique pour l'encodage de plusieurs flux simultané, let's go !

Comme je le disais, je suis en train de regarder la faisabilité d'un tel projet en générant de tel fichiers sur Github et les uploader sur un serveur personnel via FTP puisque Github pages ne permet pas la redirection avec .htaccess.
Autre problème, certains lecteurs IPTV, ne permettent pas la connexion vers certains noms de domaine ou vers des domaines non sécurisé.
Souvent le cas des domaines/hébergeurs gratuits.

Je pense notamment à IPTV smarter pro qui refuse les playlists vers Github en raw (du moins chez moi) et les playlists vers les sites en .free.fr ! (erreur de connexion, vérifiez si vous êtes connecté à Internet)

Cette proposition de solution pourrait difficilement s'appliquer, du moins je pense, aux lecteurs STB CCcam (je ne me souviens plus exactement du nom ) car ils ont un protocole particulier du peu que j'en ai vu.

Cordialement.

**Reply by BellezaEmporium** [reply]
Vu que le code source d'Xtream avait fuité il y a de cela quelques années, il est aussi possible de se faire un mini Raspberry Pi et d'y héberger le serveur Xtream. Vous effectuez le nécessaire pour filtrer les ports incriminables (25461 etc) depuis l'extérieur, et amusez vous avec la source.

**Reply by Sinok563** [reply]
Bonjour @rickeymandraque,

Ce n'est pas compliqué à faire, je l'ai fais en PHP/Mysql ;)

**Reply by rickeymandraque** [reply]
Bonjour,
Je n'ai pas été très clair effectivement, je parlais d'un cas où il serait pas possible d'utiliser LAMP.
Après, effectivement, avec une solution Xtream ou équivalent... ;)

**Reply by Sinok563** [reply]
Pardon @rickeymandraque , bonjour car je n'ai pas été poli sur mon 1er message :( . Lamp je ne connais pas du tout, je ne vais pas m'avancer. 

Comme le dis @BellezaEmporium , j'ai effectivement utilisé le script Xtream decodé dispo sur Github pour comprendre le fonctionnement .

En PHP/Mysql il m'a fallu créer 4 fichiers pour rendre compatible les flux IPTV et les EPG dispo sur iptv-org:

get.php (génère le M3U)
player_api.php (génère les json)
panel_api.php (pas trop compris a quoi il sert, car les applis IPTV que j'ai pu tester ne font pas appels a ce fichier, j'ai donc bêtement rendu compatible au besoin)
xmltv.php (génère la partie EPG)
stream.php (n'existe pas sur Xtream, ce fichier sert à tester le flux avant de le distribuer a l'utilisateur et si le flux ne fonctionne pas l'utilisateur a un message d'erreur et moi une alerte)

J'ai préféré faire un code 100% moi meme (j'aime me prendre la tete) plutôt que d'utiliser le code Xtream dispo sur Github.

Je me suis servi aussi de ce wiki https://wiki.ottrun.com/homepage/player-api-for-otr-panel

**Reply by rickeymandraque** [reply]
LAMP = Linux Apache MySQL PHP, un serveur Linux avec PHP et MySQL !
Merci pour ces infos.
Panel_api.php sert à donner les informations du compte et du serveur ainsi que tout les flux au format json à l'instar de player_api.php qui ne génère que les informations serveur et utilisateur.
En ajoutant quelques requêtes, on peut filtrer le contenu comme n'avoir que les chaînes françaises/anglaise/sportives/... Ou que les films/séries/etc ...
Pour en revenir au sujet initial, Fa9012 parle d'un décodeur et si c'est le cas et que ce n'est pas un abus de langage, cette solution n'est pas pleinement compatible je pense car le protocole utilisé doit être CCAM (je peux me tromper, je ne maîtrise plus arrivé à ce niveau parce que c'est un boooordeeel !).

**Reply by Sinok563** [reply]
😂😂😂 merci pour l'info Lamp, je passerai moins pour un con sur cette abréviation (ce n'est pas mon metier le codage , plutôt un hobbie).

Il a du se tromper en parlant de decodeur car des fois selon les pays meme Francophone nous n'avons pas les mêmes mots OU effectivement tu as raison il utilise quelques choses de différent que je ne maitrise pas du tout aussi.

En tout cas merci pour les infos que tu donnes dans ton post initial notamment sur les pages perso de chez Free.
